### DISCUSSION TITLE
behavior of insert statement with empty list is not expected

### DISCUSSION THREAD

**Discussion by liujin1993** [sqlalchemy/sqlalchemy#9645] [Usage Questions]
## question
sqlalchemy version: 2.0.9
python version: 3.10.6

Hi, when i use `session.execute(insert(Foo), data_list)` replacing `bulk_save_objects()` to bulk insert rows, I encounter a strange scenario as follow:
```python
engine = create_engine(
    'mysql+pymysql://...', 
    pool_recycle=3600, echo=True
)
Session = sessionmaker(bind=engine)
with Session() as session:
    session.bulk_save_objects([])      #  there is no generated sql
    session.execute(insert(Foo), [])   #  sqlalchemy generate a null sql: INFO sqlalchemy.engine.Engine INSERT INTO foo () VALUES ()
    session.commit()
```
![image](https://user-images.githubusercontent.com/31949674/231700610-9b398e64-406d-4309-b751-cf81c54afc37.png)

Sqlalchemy insert a null row when value is an empty list. Is it expected?

## reference
https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-insert-statements

**Comment by CaselIT** [comment]
Hi,

I don't think not running an sql when running `execute()` is a viable alternative, but I don't think we should render that statement.

When using executemany with normal sql `conn.exec_driver_sql('insert into t(value) values (?)', [])` you get an error from the db, like for sqlite the error is `(sqlite3.ProgrammingError) Incorrect number of bindings supplied.`

@zzzeek Probably the best option for this is to let the db error, instead of rewriting the insert. Do you think it would be feasible, considering cache and all? Alternatively sqlalchemy could error here, but I think letting the db do that would be better

**Reply by zzzeek** [reply]
this is a bug and this should raise well before it gets to insertmanyvalues.   IMO connection.execute(..., []) of any kind should raise.  While it's true that dbapi.executemany() will usually raise for this, at this point we have too much going on to let it get that far.   I would say the best place is in the DefaultExecutionContext where we decide that we are an "executemany".   then "if not params: raise". that's it.     I think only for exec_driver_sql(), if even that, would we leave alone, as though some driver had a way to accommodate an empty collection in executemany().   for other statement execution styles that have an executemany it would raise.

**Reply by CaselIT** [reply]
makes sense. Is there already an issue?

**Reply by zzzeek** [reply]
nope....if you want to make one that would be helpful

**Reply by CaselIT** [reply]
created https://github.com/sqlalchemy/sqlalchemy/issues/9647

**Comment by Et7f3** [comment]
I also encountered this bug today:
Hopefully my non null constraint forbid this null row to be inserted. Now I add a check if my bulk list is not empty before doing the insert.

**Comment by e3krisztian** [comment]
Also got burned by it today.

I expect it to work with an empty list, and not insert anything.
It should work with empty list - there is nothing special with empty lists.

**Reply by CaselIT** [reply]
What does "work" mean for you here?

In 2.1 a warning will be shown in this case

**Reply by e3krisztian** [reply]
I define "work" as follows in this case:

Given a `data_list` of length `n`, `session.execute(insert(Foo), data_list)` inserts exactly `n` items into the table `Foo`.

The discussion is about this is not being true for `n=0`, when an unintended insertion happens.

The expected behavior would be no insertion happening for `n=0`.

Inserting something (default values) is unexpected.
Giving a warning besides inserting something, or raising an exception  for `n=0` will not restore the expected behavior, as these are not happening for other `n` values.

----

I think raising an exception is unexpected, and a decision that forces unneeded checks in user code.

Given a `data_list` (empty or not) I want to write
```py
session.execute(insert(Foo), data_list)
```

instead of (LBYL)
```py
if data_list:
    session.execute(insert(Foo), data_list)
```

or (worse looking, the Python way EAFP)
```py
try:
    session.execute(insert(Foo), data_list)
except EmptyParameterListOrWhatever:
   pass
```

It is analogous to any of these raising an exception in Python:

- `data_list.extend(another_list)`
- `data_list += another_list`
- `1 + another_number`

when `another_list = []` and `another_number = 0`.
We expect these to be handled by the language.
These are not exceptional values.

**Reply by zzzeek** [reply]
While most DBAPIs when you pass an empty list to `.executemany()` allow it to pass silently, our `execute()` feature does lots of things a plain `.executemany()` call does not do.  Many of these things will not translate to empty lists.  

First and foremost is that the parameter lists that are formed for constructs like INSERT are *derived from the parameters themselves*.   That is, you get a different INSERT statement in each of these cases:

```py
# generates INSERT INTO table (name) VALUES (?)
connection.execute(table.insert(), [{"name": "somename"}, {"name": "someothername"}])

# generates INSERT INTO table (id, name) VALUES (?, ?)
connection.execute(table.insert(), [{"id": 1, "name": "somename"}, {"id": 2, "name": "someothername"}])
```

the Python DBAPI of course does nothing like this and for SQLAlchemy, the INSERT statement to be generated is not known if no parameters are passed.   If we were to support `.connection.execute(stmt, [])` we very much would like to be able to call down to the DBAPI `.executemany()` method with an empty list and maintain DBAPI behaviors, however in this case this is impossible.    From there it follows that `.executemany()` nor `.execute()` is ever called at all in this case, and that the `cursor` no longer has the expected state for a statement just executed; the `CursorResult` that's normally returned is therefore impossible to return in this case and overall the entire use case collapses - event hooks like `after_cursor_execute()` can now no longer be called, which is then asymmetrical with `before_cursor_execute()` which *can* still be called, although the receiver of the event hook would not have a clear picture if `.executemany` is actually to be called or not, so this would be a major compatibility issue as well.

that's the first debacle allowed by a direct allowance of empty lists, continuing on this theme we also support INSERT...RETURNING with an "executemany" call, and to achieve this we do not call "executemany" at all.   For all dialects other than Oracle we use [insertmanyvalues](https://docs.sqlalchemy.org/en/20/core/connections.html#engine-insertmanyvalues).  The INSERT statement emitted by this feature does not have any form that works with an empty set of parameters.  the statement cannot be sent to `.execute()` and therefore we cannot get an empty result set without manufacturing a fake one, which would introduce behavioral inconsistencies (such as, exceptions for invalid aspects of the INSERT statement would not be raised, `cursor.description` would be different, etc).   

I can probably think of more API degradations that would occur here by allowing this case to pass silently.   There's also the argument that we would allow a passthrough to `.executemany(stmt, [])` for those cases where we **do** have enough information to call `.executemany()` and we would in fact call this method directly, however that still leads to `.execute()` methods which would need to blow up with an empty list in some cases, pass silently in others, and nobody likes that either.

We can work on tailoring the above into an FAQ entry.  Overall SQLAlchemy's `.execute()` does a lot of things that cannot be translated to a clean `.executemany(stmt, [])` call if we dont have a parameter list.

**Reply by CaselIT** [reply]
The main issue is that it's not obvious what to do with the other bits done by execute in sqlalchemy (events, flushes, etc). Also executemany is not used automatically when a list is provided.

Overall an exception, while a bit inconvenient for users, is a lot easier to implement and maintain in SQLAlchemy

**Reply by e3krisztian** [reply]
We have 2 complementary views:

- user (library surface)
- developer (library internals)

These views have conflict on `execute`.
I accept, that it can not be meaningfully resolved with `execute` as I would like it to behave, however I maintain, that the way it is, as well as the proposed/implemented "solution" is still not OK from the user perspective.

----

What about introducing `Connection.executemany`, that would work for empty parameter list, and would not get confused by it?
The multiple parameter feature of  `Connection.execute` could be deprecated then.

(This is definitely not an original idea, as @zzzeek mentioned `executemany` above, as well it is mentioned in the docs many times, as it is part of DBAPI).

I understand the implementation gets difficult around the `INSERT RETURNING` case, where a result set is returned with columns defined by SQL, not a simple empty python list. I still think, that creating an empty result set (with no columns, as it was not created from the db output) is more consistent and understandable for the users.

**Reply by zzzeek** [reply]
you're arguing this API is flawed at its base.    Whether or not that's true, this has been SQLAlchemy's behavior for 19 years and this is the first objection that's ever been raised, I literally did not know before this discussion if the DBAPIs even supported this.   So as API flaws go we would presume the impact of this one is infinitesimal.

Your proposal to remove the list-accepting feature of `.execute()`, I am not exaggerating when I say this would impact hundreds of millions of lines of code, and to then suggest that this code would also not have a clean upgrade path, where all INSERT statements would need to written with all of the desired columns explicitly, matching the names already given in the dictionaries, would be fully untenable.

**Reply by the-vindicar** [reply]
I still don't understand why making user remember to write
```python
if data:
    await conn.execute(statement, data)
```
every time is okay, but putting a similar check inside of execute() is not okay. Are you worried that you can't distinguish between `await conn.execute(statement, [])` and `await conn.execute(statement, {})`, since the latter means "insert a line with default values"?

**Reply by zzzeek** [reply]
> I still don't understand why making user remember to write
> every time is okay, but putting a similar check inside of execute() is not okay. 

see the answer at https://github.com/sqlalchemy/sqlalchemy/discussions/9645#discussioncomment-11609881 

summary:

1. session.execute() calls into:
     a. do_orm_execute() event, which has the opportunity to modify the params
     b. various ORM -level wrappers for "execute" that make further decisions about the statement
     c. Core level execute(), which calls into:
         c1. before_execute() user event, which has the opportunity to modify the params
         c2. compile() for the statement, which relies upon the **keys** in the parameter dictionaries to compile the statement (this is why None gets inserted when no dictionaries)
         c3. transaction.autobegin(), which will BEGIN a transaction if one is not already set up
         c4. compiled context pre_exec()
         c5. DBAPI cursor.setinputsizes(), which relies upon the keys in the parameter dictionaries to know what it's doing
         c6. before_cursor_execute() user event
         c7. logging of the statement
         c8. DBAPI cursor.executemany(), which in most cases sees empty parameter list and is a no-op.  
         c9. after_cursor_execute() user event
         c10. compiled context post_exec()
         c11. creates a CursorResultSet to deliver contextual information about the cursor based on the statement which was just invoked (which was not invoked because the DBAPI skipped it)
         c12. after_execute user event
    d. ORM will in some cases do wrapping of the CursorResult with a Result, if it's a RETURNING call

There's no place a conditional like "if not parameter_list: return" could be placed without breaking expectations.    do event hooks get called or not with empty param list?   If not, event hooks that want to log statements and intercept calls are being silently disallowed from receiving these calls.   If so, then event hooks are not being called symmetrically; before_cursor_execute() but not after_cursor_execute(), because we "returned".  Or if we do call after_cursor_execute(), the state is wrong, no statement was actually invoked.

It's a huge bag of ambiguity and guessing about intent.   It will lead to problems.  So it's disallowed, so that the user application is forced to decide how it wants to deal with an empty parameter list.

**Reply by CaselIT** [reply]
it was already decided to disallow this usage, 2.1 will raise a deprecation warning, 2.2 will likely raise an error

**Reply by the-vindicar** [reply]
> 2.1 will raise a deprecation warning, 2.2 will likely raise an error

That would be an improvement over the current behaviour. Still a 'gotcha' for the user, but at least it'll be easier to understand why an error happened. Thanks.
